### ISSUE TITLE
复现结果其他数据集一致，但ETTh1差别很大

### ISSUE THREAD

**Issue by erikalien5595** [hqh0728/TimeBase#1]
作者您好，我用您的原代码尝试复现了所有数据集的结果，只有ETTh1的结果差别达到了3%，而且很奇怪的是epoch 1的损失（见下方）和您提供的log中是一致的，但是第二个epoch开始就不一致了，请问您可以解释一下产生这个现象的原因吗？

Args in experiment:
Namespace(is_training=1, model_id='ETTh1_720_96', model='LightTimeBaseTST', data='ETTh1', root_path='./dataset/', data_path='ETTh1.csv', features='M', target='OT', freq='h', checkpoints='./checkpoints/', seq_len=720, label_len=48, pred_len=96, period_len=24, basis_num=6, use_period_norm=1, use_orthogonal=1, orthogonal_weight=0.16, fc_dropout=0.05, head_dropout=0.0, patch_len=16, stride=8, padding_patch='end', revin=1, affine=0, subtract_last=0, decomposition=0, kernel_size=7, individual=0, embed_type=0, enc_in=7, dec_in=7, c_out=7, d_model=512, n_heads=8, e_layers=2, d_layers=1, d_ff=2048, moving_avg=25, factor=1, distil=True, dropout=0.05, embed='learned', activation='gelu', output_attention=False, do_predict=False, num_workers=10, itr=1, train_epochs=30, batch_size=64, patience=5, learning_rate=0.1, des='test', loss='mse', lradj='type3', pct_start=0.3, use_amp=False, use_gpu=True, gpu=2, use_multi_gpu=0, devices='0,1', test_flop=False)
Use GPU: cuda:2
214
[INFO] Register count_linear() for <class 'torch.nn.modules.linear.Linear'>.
MACs: 34272.0
Params: 214.0
34272.0 MACs
>>>>>>>start training : ETTh1_720_96_LightTimeBaseTST_ETTh1_ftM_sl720_pl96_test_0_seed2023>>>>>>>>>>>>>>>>>>>>>>>>>>
train 7825
val 2785
test 2785
	iters: 100, epoch: 1 | loss: 0.5027055
	speed: 0.0391s/iter; left time: 140.2587s
Max Memory (MB): 5.5869140625
Epoch: 1 cost time: 4.623061180114746
Epoch: 1, Steps: 123 | Train Loss: 0.4897942 Vali Loss: 0.9994107 Test Loss: 0.3918597
Validation loss decreased (inf --> 0.999411).  Saving model ...
Updating learning rate to 0.1
	iters: 100, epoch: 2 | loss: 0.3974000
	speed: 0.0733s/iter; left time: 254.3009s
Max Memory (MB): 5.94775390625
Epoch: 2 cost time: 1.6149861812591553
Epoch: 2, Steps: 123 | Train Loss: 0.4034078 Vali Loss: 0.7583007 Test Loss: 0.3997736
Validation loss decreased (0.999411 --> 0.758301).  Saving model ...
Updating learning rate to 0.1
	iters: 100, epoch: 3 | loss: 0.3943162
	speed: 0.0633s/iter; left time: 211.8035s
Max Memory (MB): 5.94775390625
Epoch: 3 cost time: 3.02715802192688
Epoch: 3, Steps: 123 | Train Loss: 0.3865392 Vali Loss: 0.7573515 Test Loss: 0.3949600
Validation loss decreased (0.758301 --> 0.757351).  Saving model ...
Updating learning rate to 0.1
	iters: 100, epoch: 4 | loss: 0.3068920
	speed: 0.0609s/iter; left time: 196.1737s
Max Memory (MB): 5.94775390625
Epoch: 4 cost time: 3.4013471603393555
Epoch: 4, Steps: 123 | Train Loss: 0.3927572 Vali Loss: 0.7765000 Test Loss: 0.3683301
EarlyStopping counter: 1 out of 5
Updating learning rate to 0.08000000000000002
	iters: 100, epoch: 5 | loss: 0.4099974
	speed: 0.1046s/iter; left time: 324.1631s
Max Memory (MB): 5.94775390625
Epoch: 5 cost time: 5.1467015743255615
Epoch: 5, Steps: 123 | Train Loss: 0.3945653 Vali Loss: 0.7457405 Test Loss: 0.3635441
Validation loss decreased (0.757351 --> 0.745741).  Saving model ...
Updating learning rate to 0.06400000000000002
	iters: 100, epoch: 6 | loss: 0.3873549
	speed: 0.1079s/iter; left time: 321.1154s
Max Memory (MB): 5.94775390625
Epoch: 6 cost time: 5.520569562911987
Epoch: 6, Steps: 123 | Train Loss: 0.3744640 Vali Loss: 0.7221637 Test Loss: 0.3623807
Validation loss decreased (0.745741 --> 0.722164).  Saving model ...
Updating learning rate to 0.051200000000000016
	iters: 100, epoch: 7 | loss: 0.3874776
	speed: 0.1082s/iter; left time: 308.6934s
Max Memory (MB): 5.94775390625
Epoch: 7 cost time: 5.501909494400024
Epoch: 7, Steps: 123 | Train Loss: 0.3685669 Vali Loss: 0.8492353 Test Loss: 0.3916617
EarlyStopping counter: 1 out of 5
Updating learning rate to 0.04096000000000001
	iters: 100, epoch: 8 | loss: 0.3690400
	speed: 0.1120s/iter; left time: 305.8616s
Max Memory (MB): 5.94775390625
Epoch: 8 cost time: 5.443071126937866
Epoch: 8, Steps: 123 | Train Loss: 0.3677843 Vali Loss: 0.7541395 Test Loss: 0.3682980
EarlyStopping counter: 2 out of 5
Updating learning rate to 0.03276800000000001
	iters: 100, epoch: 9 | loss: 0.3697669
	speed: 0.0811s/iter; left time: 211.4325s
Max Memory (MB): 5.94775390625
Epoch: 9 cost time: 3.0598556995391846
Epoch: 9, Steps: 123 | Train Loss: 0.3642505 Vali Loss: 0.7273691 Test Loss: 0.3604766
EarlyStopping counter: 3 out of 5
Updating learning rate to 0.026214400000000013
	iters: 100, epoch: 10 | loss: 0.3903659
	speed: 0.0538s/iter; left time: 133.6610s
Max Memory (MB): 5.94775390625
Epoch: 10 cost time: 2.7796552181243896
Epoch: 10, Steps: 123 | Train Loss: 0.3625220 Vali Loss: 0.7493026 Test Loss: 0.3698568
EarlyStopping counter: 4 out of 5
Updating learning rate to 0.020971520000000007
	iters: 100, epoch: 11 | loss: 0.3128429
	speed: 0.1521s/iter; left time: 359.1854s
Max Memory (MB): 5.94775390625
Epoch: 11 cost time: 6.913191556930542
Epoch: 11, Steps: 123 | Train Loss: 0.3601045 Vali Loss: 0.7850007 Test Loss: 0.3769461
EarlyStopping counter: 5 out of 5
Early stopping
Final Max Memory (MB): 5.94775390625
>>>>>>>testing : ETTh1_720_96_LightTimeBaseTST_ETTh1_ftM_sl720_pl96_test_0_seed2023<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
test 2785
mse:0.36272117495536804, mae:0.38996240496635437, rse:0.5720635652542114

**Comment by iganggang** [comment]
very good. i still can not find where the plug into patchtst.

**Comment by h505023992** [comment]
Thank you very much for your question. This issue is indeed related to the sensitivity of the dataset. The ETTh1 dataset is relatively small, and even when the model weights are fixed, slight numerical differences can arise due to variations in GPU architectures and floating-point behavior.

In our experiments, we tested both **A100** and **V100** GPUs. On the **A100**, we were able to fully reproduce the reported results (as shown in the uploaded logs). However, on the **V100**, we observed minor numerical discrepancies starting from the second epoch, which gradually accumulated and ultimately led to the model not being saved at the optimal checkpoint in terms of test-set generalization.

We appreciate your understanding and are happy to provide additional details if needed.


> 作者您好，我用您的原代码尝试复现了所有数据集的结果，只有ETTh1的结果差别达到了3%，而且很奇怪的是epoch 1的损失（见下方）和您提供的log中是一致的，但是第二个epoch开始就不一致了，请问您可以解释一下产生这个现象的原因吗？
> 
> Args in experiment: Namespace(is_training=1, model_id='ETTh1_720_96', model='LightTimeBaseTST', data='ETTh1', root_path='./dataset/', data_path='ETTh1.csv', features='M', target='OT', freq='h', checkpoints='./checkpoints/', seq_len=720, label_len=48, pred_len=96, period_len=24, basis_num=6, use_period_norm=1, use_orthogonal=1, orthogonal_weight=0.16, fc_dropout=0.05, head_dropout=0.0, patch_len=16, stride=8, padding_patch='end', revin=1, affine=0, subtract_last=0, decomposition=0, kernel_size=7, individual=0, embed_type=0, enc_in=7, dec_in=7, c_out=7, d_model=512, n_heads=8, e_layers=2, d_layers=1, d_ff=2048, moving_avg=25, factor=1, distil=True, dropout=0.05, embed='learned', activation='gelu', output_attention=False, do_predict=False, num_workers=10, itr=1, train_epochs=30, batch_size=64, patience=5, learning_rate=0.1, des='test', loss='mse', lradj='type3', pct_start=0.3, use_amp=False, use_gpu=True, gpu=2, use_multi_gpu=0, devices='0,1', test_flop=False) Use GPU: cuda:2 214 [INFO] Register count_linear() for <class 'torch.nn.modules.linear.Linear'>. MACs: 34272.0 Params: 214.0 34272.0 MACs
> 
> > > > > > > > start training : ETTh1_720_96_LightTimeBaseTST_ETTh1_ftM_sl720_pl96_test_0_seed2023>>>>>>>>>>>>>>>>>>>>>>>>>>
> > > > > > > > train 7825
> > > > > > > > val 2785
> > > > > > > > test 2785
> > > > > > > > iters: 100, epoch: 1 | loss: 0.5027055
> > > > > > > > speed: 0.0391s/iter; left time: 140.2587s
> > > > > > > > Max Memory (MB): 5.5869140625
> > > > > > > > Epoch: 1 cost time: 4.623061180114746
> > > > > > > > Epoch: 1, Steps: 123 | Train Loss: 0.4897942 Vali Loss: 0.9994107 Test Loss: 0.3918597
> > > > > > > > Validation loss decreased (inf --> 0.999411).  Saving model ...
> > > > > > > > Updating learning rate to 0.1
> > > > > > > > iters: 100, epoch: 2 | loss: 0.3974000
> > > > > > > > speed: 0.0733s/iter; left time: 254.3009s
> > > > > > > > Max Memory (MB): 5.94775390625
> > > > > > > > Epoch: 2 cost time: 1.6149861812591553
> > > > > > > > Epoch: 2, Steps: 123 | Train Loss: 0.4034078 Vali Loss: 0.7583007 Test Loss: 0.3997736
> > > > > > > > Validation loss decreased (0.999411 --> 0.758301).  Saving model ...
> > > > > > > > Updating learning rate to 0.1
> > > > > > > > iters: 100, epoch: 3 | loss: 0.3943162
> > > > > > > > speed: 0.0633s/iter; left time: 211.8035s
> > > > > > > > Max Memory (MB): 5.94775390625
> > > > > > > > Epoch: 3 cost time: 3.02715802192688
> > > > > > > > Epoch: 3, Steps: 123 | Train Loss: 0.3865392 Vali Loss: 0.7573515 Test Loss: 0.3949600
> > > > > > > > Validation loss decreased (0.758301 --> 0.757351).  Saving model ...
> > > > > > > > Updating learning rate to 0.1
> > > > > > > > iters: 100, epoch: 4 | loss: 0.3068920
> > > > > > > > speed: 0.0609s/iter; left time: 196.1737s
> > > > > > > > Max Memory (MB): 5.94775390625
> > > > > > > > Epoch: 4 cost time: 3.4013471603393555
> > > > > > > > Epoch: 4, Steps: 123 | Train Loss: 0.3927572 Vali Loss: 0.7765000 Test Loss: 0.3683301
> > > > > > > > EarlyStopping counter: 1 out of 5
> > > > > > > > Updating learning rate to 0.08000000000000002
> > > > > > > > iters: 100, epoch: 5 | loss: 0.4099974
> > > > > > > > speed: 0.1046s/iter; left time: 324.1631s
> > > > > > > > Max Memory (MB): 5.94775390625
> > > > > > > > Epoch: 5 cost time: 5.1467015743255615
> > > > > > > > Epoch: 5, Steps: 123 | Train Loss: 0.3945653 Vali Loss: 0.7457405 Test Loss: 0.3635441
> > > > > > > > Validation loss decreased (0.757351 --> 0.745741).  Saving model ...
> > > > > > > > Updating learning rate to 0.06400000000000002
> > > > > > > > iters: 100, epoch: 6 | loss: 0.3873549
> > > > > > > > speed: 0.1079s/iter; left time: 321.1154s
> > > > > > > > Max Memory (MB): 5.94775390625
> > > > > > > > Epoch: 6 cost time: 5.520569562911987
> > > > > > > > Epoch: 6, Steps: 123 | Train Loss: 0.3744640 Vali Loss: 0.7221637 Test Loss: 0.3623807
> > > > > > > > Validation loss decreased (0.745741 --> 0.722164).  Saving model ...
> > > > > > > > Updating learning rate to 0.051200000000000016
> > > > > > > > iters: 100, epoch: 7 | loss: 0.3874776
> > > > > > > > speed: 0.1082s/iter; left time: 308.6934s
> > > > > > > > Max Memory (MB): 5.94775390625
> > > > > > > > Epoch: 7 cost time: 5.501909494400024
> > > > > > > > Epoch: 7, Steps: 123 | Train Loss: 0.3685669 Vali Loss: 0.8492353 Test Loss: 0.3916617
> > > > > > > > EarlyStopping counter: 1 out of 5
> > > > > > > > Updating learning rate to 0.04096000000000001
> > > > > > > > iters: 100, epoch: 8 | loss: 0.3690400
> > > > > > > > speed: 0.1120s/iter; left time: 305.8616s
> > > > > > > > Max Memory (MB): 5.94775390625
> > > > > > > > Epoch: 8 cost time: 5.443071126937866
> > > > > > > > Epoch: 8, Steps: 123 | Train Loss: 0.3677843 Vali Loss: 0.7541395 Test Loss: 0.3682980
> > > > > > > > EarlyStopping counter: 2 out of 5
> > > > > > > > Updating learning rate to 0.03276800000000001
> > > > > > > > iters: 100, epoch: 9 | loss: 0.3697669
> > > > > > > > speed: 0.0811s/iter; left time: 211.4325s
> > > > > > > > Max Memory (MB): 5.94775390625
> > > > > > > > Epoch: 9 cost time: 3.0598556995391846
> > > > > > > > Epoch: 9, Steps: 123 | Train Loss: 0.3642505 Vali Loss: 0.7273691 Test Loss: 0.3604766
> > > > > > > > EarlyStopping counter: 3 out of 5
> > > > > > > > Updating learning rate to 0.026214400000000013
> > > > > > > > iters: 100, epoch: 10 | loss: 0.3903659
> > > > > > > > speed: 0.0538s/iter; left time: 133.6610s
> > > > > > > > Max Memory (MB): 5.94775390625
> > > > > > > > Epoch: 10 cost time: 2.7796552181243896
> > > > > > > > Epoch: 10, Steps: 123 | Train Loss: 0.3625220 Vali Loss: 0.7493026 Test Loss: 0.3698568
> > > > > > > > EarlyStopping counter: 4 out of 5
> > > > > > > > Updating learning rate to 0.020971520000000007
> > > > > > > > iters: 100, epoch: 11 | loss: 0.3128429
> > > > > > > > speed: 0.1521s/iter; left time: 359.1854s
> > > > > > > > Max Memory (MB): 5.94775390625
> > > > > > > > Epoch: 11 cost time: 6.913191556930542
> > > > > > > > Epoch: 11, Steps: 123 | Train Loss: 0.3601045 Vali Loss: 0.7850007 Test Loss: 0.3769461
> > > > > > > > EarlyStopping counter: 5 out of 5
> > > > > > > > Early stopping
> > > > > > > > Final Max Memory (MB): 5.94775390625
> > > > > > > > testing : ETTh1_720_96_LightTimeBaseTST_ETTh1_ftM_sl720_pl96_test_0_seed2023<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
> > > > > > > > test 2785
> > > > > > > > mse:0.36272117495536804, mae:0.38996240496635437, rse:0.5720635652542114

**Comment by h505023992** [comment]
We apologize for the inconvenience caused by not clearly annotating this earlier. In the repository
[TimeBase](https://github.com/hqh0728/TimeBase/tree/main)/[plug-and-play_for_patchtst](https://github.com/hqh0728/TimeBase/tree/main/plug-and-play_for_patchtst)/[layers](https://github.com/hqh0728/TimeBase/tree/main/plug-and-play_for_patchtst/layers)

`PatchTST_backbone.py`, **line 157**, the line

```python
self.ts2basis = nn.Linear(self.patch_num, configs.basis_num)
```

implements the *plug-in mechanism* for integrating our method into PatchTST.

Thank you for your understanding, and please feel free to let us know if anything remains unclear. @iganggang  



> very good. i still can not find where the plug into patchtst.

**Comment by erikalien5595** [comment]
My question has been resolved. Thank you for your response。

> 非常感谢您的提问。这个问题确实与数据集的敏感性有关。ETTh1 数据集相对较小，即使模型权重固定，由于 GPU 架构和浮点运算行为的差异，也可能出现细微的数值偏差。
> 
> 在我们的实验中，我们测试了**A100**和**V100**两种GPU。在**A100**上，我们能够完全复现报告的结果（如上传的日志所示）。然而，在**V100**上，我们观察到从第二个epoch开始出现一些细微的数值偏差，这些偏差逐渐累积，最终导致模型在测试集泛化能力方面未能保存到最佳检查点。
> 
> 感谢您的理解，如有需要，我们很乐意提供更多详情。
